php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
         ]
     ];
 
-    public static function isCurrentLevelComplete($Student, $Competency)
+    public static $dynamicFields = [
+        'demonstrationsLogged' => [
+            'getter' => 'getDemonstrationsLogged'
+        ],
+        'demonstrationsComplete' => [
+            'getter' => 'getDemonstrationsComplete'
+        ],
+        'demonstrationsAverage' => [
+            'getter' => 'getDemonstrationsAverage'
+        ],
+        'demonstrationData' => [
+            'getter' => 'getDemonstrationData'
+        ],
+        'effectiveDemonstrationsData' => [
+            'getter' => 'getEffectiveDemonstrationsData'
+        ]
+
+    ];
+
+    public function calculateStartingRating($autoSave = true)
     {
-        $completion = $Competency->getCompletionForStudent($Student);
+        $validSkills = DB::allRecords(
+            'SELECT ID, '.
+            ' CAST( '.
+                'REPLACE( '.
+            '       IFNULL ( '.
+            '           JSON_EXTRACT(DemonstrationsRequired, CONCAT(\'$."\', %u, \'"\')), '.
+            '           JSON_EXTRACT(DemonstrationsRequired, "$.default")'.
+            '       ), '.
+            '   \'"\', ' .
+            '   "")'.
+            '   AS UNSIGNED '.
+            ' ) AS DemonstrationsRequirements '.
+            '  FROM `%s` '.
+            ' WHERE CompetencyID = %u '.
+            'HAVING DemonstrationsRequirements > 0 ',
+            [
+                $this->Level,
+                Skill::$tableName,
+                $this->Competency->ID
+            ]
+        );
+        $validSkillIds = array_column($validSkills, 'ID');
```

Comment:
@nbey let's just foreach through `$this->Competency->Skills` and have that build up `$validSkillsId`, we're not gaining any efficiency doing all that json stuff in MySQL